monash-human-power/MHP_DAS
# Add MQTT radio bridge

## Description

*begin readme copy-paste*

This program is used to join (or "bridge") the two MQTT brokers on the bike and the chase car so that MQTT messages may be sent without relying on being on the same local network.

At a high level, this works by making a subscription to all MQTT topics (`#`), and then for each received message:

- Check the message was not one that was sent by the bridge itself (performed in the `Bridge` class),
- Serialise the message into a series of bytes (performed in the `TxProtocol` class),
- Pack the bytes into one or more `Frame`s, or radio packets (also `TxProtocol`),
- Send each of the `Frame`s over the radio (using `ZetaRfRadio`).

On the other end of the bridge, the reverse process would happen.

- A packet is received by `ZetaRfRadio`,
- It is processed by `RxProtocol`, which will emit create a MQTT message when enough packets have arrived,
- `Bridge` will register the message as one sent by the MQTT radio bridge, before finally,
- The MQTT message is published using `MqttBridgeClient`.

Details on the protocol used for converting an MQTT message into bytes, and then into packets may be found on Notion.

*end readme copy-paste*

### Notes for reviewing

This is more of an MVP than a finished product. It works, but it has some notable flaws which must be fixed before we want to consider this remotely complete.
- If both ends of the bridge transmit a packet simultaneously, neither packet is successfully received. We'll likely need to add some sort of timing to our protocol to allow the ends of the bridge to take turns transmitting.
- There are no command line arguments or config files. The MQTT broker address is hardcoded to `tcp://localhost:1883` inside `main.cpp` if you wish to change it.
- Much of the code (the protocol stuff and the data structures) would be great candidates for unit tests. We should write some. Might need a third partly library? Need to research testing C++.
- I feel like there's something else I've forgotten.

I'm aware of the above problems and will save them for a future PR so we at least have an MVP merged for now. Other than that though, let me know of any and every suggestion/question you have!

If you're unfamiliar with C/C++ code: the `.hpp` files are "headers" and generally contain like an outline of the classes, with the names/signatures of their functions and their member variables. The `.cpp` files are where the functions are actually implemented and where all the logic happens. I'd start looking at the headers (which is also where most of the documentation is) to get an idea of what the class does, and then look at how it works in the `.cpp` afterwards.

Finally, I apologise for the size of the PR, I should have broken it up more, but I didn't realise the scope of the task when I started. I'll point out that the stuff in `Bridge/src/rf_configs` can be mostly ignored, especially the `.xml` file, as this is just configuration for the si4455 chip which is mostly auto-generated. (A bit more info on that in the readme.) Also, a lot of the code is documentation. I hope it makes sense and isn't too huge, let me know if anything doesn't makes sense or if you need more of a code tour.

## Screenshots

Receiving a hello world MQTT message in a debug build:

![image](https://user-images.githubusercontent.com/17876556/89733374-6a358d00-da98-11ea-92a7-facde2b83013.png)

## Steps to Test

If you have a raspberry pi, you should be able to follow the steps in the readme.

If not, you may be able to at least compile and debug most of the code on your computer, excluding the Zeta radio section. I'm just guessing what should work so let me know if it doesn't.

- Install the Microsoft's CMake Tools vscode extension, and the C/C++ extension.
- Comment out (using `#`) the "Build/configure ZetaRF" section of `CMakeLists.txt`, as well as the line `"src/zeta.cpp"`
- Comment out (using `//`) lines 5 and 26 of `bridge.hpp`, and lines 10 and 21 of `bridge.cpp`. In other words, comment out all mention of the zeta radio.
- You should hopefully then be able to build/run/debug using the buttons on the vscode status bar.
